### ISSUE TITLE
[FlexAttention] get poor performance with Triton based GEMM kernel.

### ISSUE THREAD

**Issue by liangan1** [intel/intel-xpu-backend-for-triton#3518]
### Describe the issue

We ever validated prefill stage of LLama-3.1-8B(head_dim=128) and can get the expected performance with the advance path. But we can not get the on-par performance with the default path and the performance is too poor to accept. if you want to reproduce the performance, you can refer to 

- Pytorch commit

    https://github.com/pytorch/pytorch/pull/143553

- Triton commit

  https://github.com/intel/intel-xpu-backend-for-triton
  cd ~/intel-xpu-backend-for-triton/python/
  python setup.py install
 

- How to run model

  Transformer codebse:
  git clone [https://github.com/blzheng/transformers -b beilei/enable_flex_attn](https://github.com/blzheng/transformers%20-b%20beilei/enable_flex_attn)
  cd transformers && pip install -e ./
  Dowonload the example prompt file: 
  wget https://intel-extension-for-pytorch.s3.amazonaws.com/miscellaneous/llm/prompt.json
         reference launch script:         
            wget https://raw.githubusercontent.com/liangan1/LLM-Examples/refs/heads/main/run_llm_inductor_greedy.py

### Environment details

main branch.

**Comment by liangan1** [comment]
@whitneywhtsang  is there any updates about this issue?

**Comment by whitneywhtsang** [comment]
> [@whitneywhtsang](https://github.com/whitneywhtsang) is there any updates about this issue?

We are able to reproduce the performance observations, we will aim to improve the performance on default path, instead of advanced path.

**Comment by vlad-penkin** [comment]
Duplicate for: 
- #3486

**Comment by chengjunlu** [comment]
Reopen this issue as the #3486 is going to be not supported.

**Comment by mfrancepillois** [comment]
@liangan1 @hoshibara I noticed a few compiler options missing when launching triton kernel for XPU.
When using the default path for PVC targets, you need to set grf_mode='large' and one_matrix_per_load_for_bt: true.
Indeed, a dirty patch adding the following lines:

"grf_mode": 'large',
"one_matrix_per_load_for_bt": True,

in this file : https://github.com/liangan1/pytorch/blob/liangan1/flex_attention/torch/_inductor/runtime/triton_heuristics.py#L520-L525
has shown better performance.

**Comment by hoshibara** [comment]
In the prefill stage with 1024-token inputs, The compiled SDPA (triton_tem_fused_add_cat_mul_7) is 4.6× slower than xetla's SDPA kernel.
This shows a large performance gap.

The following hyperparameters have been tuned and set to:
- BLOCK_M: 16
- BLOCK_N: 16
- num_wraps: 4
- num_stages: 1

Model: `meta-llama/Meta-Llama-3.1-8B`

Applied this patch: https://github.com/intel/intel-xpu-backend-for-triton/issues/3486#issuecomment-2734062876
```python
options['grf_mode'] = 'large'
options['one_matrix_per_load_for_bt'] = True
```

**Comment by hoshibara** [comment]
After fixing the accuracy issue of flex decoding, we tested the performance of flex decoding.
In the test, the SDPA memory reading speed in the decoding stage is only around 50GB/s.
Could you help us confirm this case?

Model: `meta-llama/Meta-Llama-3.1-8B`

Applied this patch: https://github.com/intel/intel-xpu-backend-for-triton/issues/3486#issuecomment-2734062876
```python
options['grf_mode'] = 'large'
options['one_matrix_per_load_for_bt'] = True
```

**Comment by liangan1** [comment]
@whitneywhtsang  The FlexAttention performance for both flexattention and flexdecoding kernel is still far behind our expectation, to meet the torch-2.8 target, can you help to prioritize it?

**Comment by etiotto** [comment]
So what is the performance update on this work item ? From reading this ticket is not clear to me whether the original description here https://github.com/intel/intel-xpu-backend-for-triton/issues/3518#issue-2876542857 is relevant. IS the advanced path performance better than the default path. If so where, which kernel ?

**Comment by etiotto** [comment]
@chengjunlu have you been able to derive concrete work items that would (1) clearly define what is the performance gap (including the exact kernel it applies to) and (2) the work items that are required to investigate and/or fix the performance issues ?

**Comment by LiyangLingIntel** [comment]
Reproduced the performance of [sglang/extend_attention](https://github.com/liangan1/sglang/blob/main/python/sglang/srt/layers/extend_attention.py), which was the attention kernel of llama3.1-8 model.
With the Triton(3.3.0+gitded75067) + PyTorch(2.8.0a0+gita944833), on PVC 1100, it got close kernel time on Advanced path and default path.
* Advanced path: 0.34 ms
* Default path: 0.33 ms
The current torch FlexAttention implementation is different than the sglang extend_attention, and it's irreproducable on Advanced path. I think the performance gap should come from the implemetation difference, which Triton optimization do not cover.

**Comment by liangan1** [comment]
@LiyangLingIntel is there any update about performance?

**Comment by leonling-ll** [comment]
Merged https://github.com/intel/intel-xpu-backend-for-triton/pull/4354 for removing conversion layout for store. It can get ~20% performance improvements. Following charts show current flex attention kernels status, see the blue line in the first chart and orange line in the second chart.

![Image](https://github.com/user-attachments/assets/6943b15a-253b-4673-83ed-1df444be4d7f)

![Image](https://github.com/user-attachments/assets/21016a9e-6fb1-4c39-afbf-7d983c540dc8)

Next will be validating and integrating IGC optimization based on previous discussion, tracked by https://github.com/intel/intel-xpu-backend-for-triton/issues/4405.

**Comment by hoshibara** [comment]
I updated the llama3.1 benchmark script to be compatible with the official transformers repo.
https://github.com/hoshibara/flex_attention-pr-check

Before running this script, transformers need to be patched by:
```
git apply ../transformers-patch-for-timing.diff
```

**Comment by chengjunlu** [comment]
Performance update for llama3.1 8B model on PVC 1100 and NV A100 PCIE 40G:
xpu:
 
---------- Summary: ----------
inference-latency: 0.345 sec.
first-token-latency: 0.094 sec.
rest-token-latency: 0.027 sec.
P90-rest-token-latency: 0.027 sec.
 
 
cuda:
 
---------- Summary: ----------
inference-latency: 0.280 sec.
first-token-latency: 0.110 sec.
rest-token-latency: 0.018 sec.
P90-rest-token-latency: 0.019 sec.

**Comment by liangan1** [comment]
The first token performance is cool. do you have any comments about about the next token?

**Comment by chengjunlu** [comment]
> The first token performance is cool. do you have any comments about about the next token?

We are optimizing the flex decoding performance in micro-benchmark. And then need to align the performance to the next token in E2E model running.

**Comment by chengjunlu** [comment]
@liangan1 Can we close this issue since the flex attn of XPU has been merged to PT 2.9 ? 
Let's create the new issue to track the performance optimization further.

**Comment by liangan1** [comment]
@chengjunlu  we are working on the performance analysis, let us decide later according to the breakdown result.
